src/main/java/com/google/flourbot/entity/action/Action.java

```diff
@@ -1,7 +1,7 @@
 package com.google.flourbot.entity.action;
 
-public abstract class Action {
-  public abstract ActionType getActionType();
-  public abstract String getSheetId();
-  public abstract String getSheetUrl();
+public interface Action {
```

Comment:
Just out of curiosity, why did we switch it from abstract class to an interface?

Reply:
Good question! I want to see what other people say, but my reasoning was:

- Interfaces cannot have implementation of the methods, so we would want SheetAction to be an abstract class for sure. This doesn't tell us anything about Action yet.

- Since SheetAction is an abstract class, and SheetAppendRowAction extends it, SheetAppendRowAction therefore inherits the (non-abstract) methods that were implemented in SheetAction, and has to implement the abstract methods of SheetAction.
 - Also, in an abstract class, we can have private and protected members, which we can't have in an interface.

- Action doesn't really have anything (no functionality) that we want SheetAction to inherit, instead, it provides the method signatures that we need to implement and nothing else. Action simply provides a common structure that classes need to follow. Which to make makes more sense to have as an interface

Reply:
Ok, thanks! And ya that's a good idea to use protected for things like the sheet url.

Reply:
Interesting discussion. In my mind, Interface is more applicable when we have multiple implementations that have the same functionality but fulfilled in a different way. Like implementation that talks to Drive and Office365 to operate on a spreadsheet. Abstract class is more applicable when you have a hierarchy of object types in which each level of the hierarchy is more specific than the parent. So here abstract class makes more sense to me here...

Reply:
Changed it!